Remove experimental from gateway infrastructure

keithmattix opened the pull request:
**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind gep
/kind test

Optionally add one or more of the following kinds if applicable:
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/area conformance
-->
/kind gep
/kind cleanup
**What this PR does / why we need it**:
Removes experimental from gateway infrastructure type.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #1867

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
Promoted GEP-1867 Per Gateway Infrastructure to Standard Channel
```

youngnick commented:
#3197 actually probably blocks this to some extent, because we need to resolve that before we release. Doesn't stop us merging this, but we need to make sure we do both.

keithmattix commented:
Thanks for the reminder! @snorwin happy to wait on your change or make add the new LabelValue field in this PR. Up to you 🙂

snorwin commented:
@keithmattix I created a proposal in #3284. It would be great if you have a look at it too.

robscott commented:
@keithmattix @howardjohn We got a good question about this GEP at SIG-Network yesterday. How do we handle mutability? If these labels were to map to Pods, it could be rather disruptive to update labels.

keithmattix commented:
In the Istio implementation, we create a deployment for the gateway, so any pod restarts would be via a rolling restart. I think as long as you're not changing a label used for deployment or service selection, you're probably fine in most cases. I think the GEP reserves the gateway.networking.k8s.io prefix (i.e. users can't set labels with that prefix), and that's what we use for deployment + service selection in Istio.

robscott commented:
@keithmattix do you mind adding a note somewhere in the GEP as part of this PR to at least call out the risk that changing labels could lead to underlying dataplane Pods being replaced, depending on the underlying implementation? I'm assuming that at least some Gateway implementations may not be able to do that without at least some minor disruptions to traffic.

robscott commented on apis/v1/gateway_types.go:
```diff
@@ -226,7 +226,6 @@ type GatewaySpec struct {
 	//
 	// Support: Core
```
```suggestion
	// Support: Extended
```

I think we can theoretically call this "extended" because we have some conformance tests that will cover a subset of implementations. With that said, it's possible that a cloud-based solution may both support this feature and not pass those tests. That part starts to make it feel a bit closer to "implementation-specific", but will defer to you on which you think is more accurate here. In either case, I'm confident that we can't call this (or any of the fields within infra) "Core".

keithmattix replied:
Ah right, not sure what I was thinking here. Extended is what I was going for, with individual fields perhaps being implementation specific.

keithmattix commented:
@robscott I added some language about dataplane pods; let me know if that's sufficient. I also removed references to gatewayclass.infrastructure because I believe that direction was abandoned

robscott commented on geps/gep-1867/index.md:
```diff
@@ -75,43 +75,74 @@ For example, if I wanted to represent a `version` field and change that to trigg
-  // fields not expressed directly in this struct.
-  // This follows the same semantics as GatewayClass's ParametersRef, but lives on the Gateway.
-  ParametersRef ParametersReference
+  // Labels that SHOULD be applied to any resources created in response to this Gateway.
+	//
```
Tiny nit: Looks like a combination of tabs and spaces happening here and elsewhere in md file now.

robscott commented on geps/gep-1867/index.md:
````diff
@@ -75,43 +75,74 @@ For example, if I wanted to represent a `version` field and change that to trigg
 ```
 
+!!! warning
+    Modifying pod labels via `infrastructure.labels` could result in data plane pods being replaced (depending on the implementation). The way in which the pods are replaced is implementation specific and may result in downtime. Implementations should document how they handle changes to `infrastructure.labels`.
````
I think something like this could likely belong in the spec for infrastructure.labels as well. Here's a rough starting point, feel free to change:

"If an implementation maps these labels to Pods, or any other resource that would need to be recreated when labels change, it SHOULD clearly warn about this behavior in documentation."

robscott replied:
@keithmattix not sure if you intended to add this to spec, but I actually meant that it could be useful to add a comment like that to the spec somewhere around here: https://github.com/kubernetes-sigs/gateway-api/blob/24f73922acf501b547e0f183c8f8066eb6fbb045/apis/v1/gateway_types.go#L707-L720

keithmattix replied:
Ah good catch; I had the markdown open on the API section, and my eyes told my brain I was editing a Go file 🤦🏾 Should be fixed no!

robscott reviewed (commented):
Thanks @keithmattix!

robscott commented:
Thanks @keithmattix! LGTM once presubmits pass - looks like another round of codegen is needed.

/approve

youngnick commented:
/lgtm

once the presubmits pass. Nice work @keithmattix!

keithmattix commented on config/crd/experimental/gateway.networking.k8s.io_gateways.yaml:
```diff
@@ -148,13 +148,11 @@ spec:
                 type: string
               infrastructure:
-                description: |+
+                description: |-
```
@robscott @youngnick I think this is the reason codegen is failing? Not sure what this indicates

robscott replied:
🤔 maybe it's tied to go version? Or maybe k8s deps? I think I've seen similar changes recently in other PRs, but wasn't sure what was leading to it

youngnick replied:
According to https://yaml.org/spec/1.2-old/spec.html#id2794534, we're changing from the `+` chomping operator to the `-` chomping operator, which will change the behavior of the literal string such that the final line break and trailing empty lines will be removed (`-`) instead of kept (`+`).

I would guess that the tooling in a specific Go version has changed this somewhere. I've had problems with codegen like this before where my Go version and modules didn't exactly match the ones in the repo.

Personally, I have multiple Go versions installed using https://go.dev/doc/manage-install, and then use the `direnv` tool to have an `.envrc` file that selects the correct version when in that directory, which has seeming stopped this sort of issue happening as much for me. (Visual Studio Code also needs to be poked with the correct version.)

keithmattix replied:
Oh actually it's not this file at all; it's the standard version

keithmattix commented:
Oh the cel isn't getting copied over hmmm

keithmattix commented:
It was because I didn't rebase on top of #3197 🤦🏾 presubmit should pass now

robscott commented:
> It was because I didn't rebase on top of #3197 🤦🏾 presubmit should pass now

Good catch, thanks @keithmattix!

/lgtm

keithmattix commented:
/hold cancel
